pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -144,10 +207,79 @@ func GroupHibernationSchedulesByLocation(schedules []gardencorev1beta1.Hibernati
 			locationID = time.UTC.String()
 		}
 
-		locationToSchedules[locationID] = append(locationToSchedules[locationID], schedule)
+		location, err := time.LoadLocation(locationID)
+		if err != nil {
+			return nil, err
+		}
+		if schedule.Start != nil {
+			parsed, err := cron.ParseStandard(*schedule.Start)
+			if err != nil {
+				return nil, err
+			}
+			parsedHibernationSchedules = append(parsedHibernationSchedules,
+				ParsedHibernationSchedule{Location: location, Schedule: parsed, Enabled: true},
+			)
+		}
+		if schedule.End != nil {
+			parsed, err := cron.ParseStandard(*schedule.End)
+			if err != nil {
+				return nil, err
+			}
+			parsedHibernationSchedules = append(parsedHibernationSchedules,
+				ParsedHibernationSchedule{Location: location, Schedule: parsed, Enabled: false},
+			)
+		}
+	}
+
+	return parsedHibernationSchedules, nil
+}
+
+// nextHibernationTimeDuration returns the time duration after which to requeue the shoot based on the hibernation schedules and current time.
+func nextHibernationTimeDuration(schedules []ParsedHibernationSchedule, now time.Time) time.Duration {
+	var timestamps []time.Time
+	for _, schedule := range schedules {
+		ts := schedule.Next(now)
+		timestamps = append(timestamps, ts)
+	}
+
+	sort.Slice(timestamps, func(i, j int) bool {
+		return timestamps[i].Before(timestamps[j])
+	})
+
+	duration := timestamps[0].Sub(now)
+	return duration
+}
+
+// getScheduleWithMostRecentTime returns the ParsedHibernationSchedule that contains the schedule with the most recent previous execution time.
+func getScheduleWithMostRecentTime(schedules []ParsedHibernationSchedule, now time.Time, shoot *gardencorev1beta1.Shoot) *ParsedHibernationSchedule {
+	var startTime time.Time
+	if shoot.Status.LastHibernationTriggerTime != nil {
+		startTime = shoot.Status.LastHibernationTriggerTime.Time
+	} else {
+		// if the shoot has just been created or has never been hibernated, use the creation timestamp
+		startTime = shoot.CreationTimestamp.Time
+	}
```

Comment:
```suggestion
	lastHibernationTriggerTime := shoot.CreationTimestamp.Time // if the shoot has just been created or has never been hibernated, use the creation timestamp
	if shoot.Status.LastHibernationTriggerTime != nil {
		lastHibernationTriggerTime = shoot.Status.LastHibernationTriggerTime.Time
	}
```

However, as discussed earlier, I think it's a good idea to cap this. Probably we can just use `7d` here?